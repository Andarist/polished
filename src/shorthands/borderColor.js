// @flow
import { _directionalProperty } from '../helpers/directionalProperty'
import polish from '../validation/polish'

/**
 * Shorthand that accepts up to four values, including null to skip a value, and maps them to their respective directions.
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...borderColor('red', 'green', 'blue', 'yellow')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${borderColor('red', 'green', 'blue', 'yellow')}
 * `
 *
 * // CSS as JS Output
 *
 * div {
 *   'borderTopColor': 'red',
 *   'borderRightColor': 'green',
 *   'borderBottomColor': 'blue',
 *   'borderLeftColor': 'yellow'
 * }
 */

function borderColor(...values: Array<?string>) {
  return _directionalProperty('borderColor', ...values)
}

// TODO: Proper Color Type
export default polish({
  modulePath: 'shorthands/borderColor',
  types: [
    { key: 'firstColor', type: 'string' },
    { key: 'secondColor', type: 'string' },
    { key: 'thirdColor', type: 'string' },
    { key: 'fourthColor', type: 'string' },
  ],
})(borderColor)
