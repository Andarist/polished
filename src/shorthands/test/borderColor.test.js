// @flow
import borderColor from '../borderColor'

describe('borderColor', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('properly applies a value when passed only one', () => {
    expect(borderColor('red')).toMatchSnapshot()
  })
  it('properly applies values when passed two', () => {
    expect(borderColor('red', 'blue')).toMatchSnapshot()
  })
  it('properly applies values when passed three', () => {
    expect(borderColor('red', 'blue', 'green')).toMatchSnapshot()
  })
  it('properly applies values when passed four', () => {
    expect(borderColor('red', 'blue', 'green', 'yellow')).toMatchSnapshot()
  })
  it('throws an error for each invalid value', () => {
    borderColor(1, 2, 3, 4)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
  it('throws a warning when passed too many parameters', () => {
    borderColor('red', 'blue', 'green', 'yellow', 'purple')
    // eslint-disable-next-line no-console
    expect(console.warn.mock.calls).toMatchSnapshot()
  })
})
