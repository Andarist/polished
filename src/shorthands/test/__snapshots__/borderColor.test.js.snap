exports[`borderColor properly applies a value when passed only one 1`] = `
Object {
  "borderBottomColor": "red",
  "borderLeftColor": "red",
  "borderRightColor": "red",
  "borderTopColor": "red",
}
`;

exports[`borderColor properly applies values when passed four 1`] = `
Object {
  "borderBottomColor": "green",
  "borderLeftColor": "yellow",
  "borderRightColor": "blue",
  "borderTopColor": "red",
}
`;

exports[`borderColor properly applies values when passed three 1`] = `
Object {
  "borderBottomColor": "green",
  "borderLeftColor": "blue",
  "borderRightColor": "blue",
  "borderTopColor": "red",
}
`;

exports[`borderColor properly applies values when passed two 1`] = `
Object {
  "borderBottomColor": "red",
  "borderLeftColor": "blue",
  "borderRightColor": "blue",
  "borderTopColor": "red",
}
`;

exports[`borderColor throws a warning when passed too many parameters 1`] = `
Array [
  Array [
    "%c ✨ WARNING ✨ ---- shorthands/borderColor.js --%c

borderColor %cexpects a maximum of %c4 parameters%c. However, you passed %c5 parameters%c. This additional parameter was ignored.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderColor%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: goldenrod",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`borderColor throws an error for each invalid value 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderColor.js --%c

borderColor %cexpects a %c1st parameter%c(%cfirstColor%c: %cstring%c). However, you passed (%c\'1\'%c:%cnumber%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderColor%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderColor.js --%c

borderColor %cexpects a %c2nd parameter%c(%csecondColor%c: %cstring%c). However, you passed (%c\'2\'%c:%cnumber%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderColor%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderColor.js --%c

borderColor %cexpects a %c3rd parameter%c(%cthirdColor%c: %cstring%c). However, you passed (%c\'3\'%c:%cnumber%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderColor%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- shorthands/borderColor.js --%c

borderColor %cexpects a %c4th parameter%c(%cfourthColor%c: %cstring%c). However, you passed (%c\'4\'%c:%cnumber%c) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#borderColor%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: green; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: red; font-size: 12px; font-style: italic",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
